flamedisx/likelihood.py

```diff
@@ -371,166 +375,162 @@ def bestfit(self,
-        :param kind: 'upper', 'lower' or 'central' type of limit/interval
+        If omitted, will compute it using bestfit.
+        :param guess: {param: value} guess of the result, or None.
+        If omitted, nuisance parameters will be guessed equal to bestfit.
+        If omitted, guess for target parameters will be based on asymptotic
+        parabolic computation.
+        :param fix: {param: value} to fix during interval computation.
+        Result is only valid if same parameters were fixed for bestfit.
+        :param confidence_level: Requried confidence level of the interval
+        :param kind: Type of interval, 'upper', 'lower' or 'central'
+        :param sigma_guess: Guess for one sigma uncertainty on the target
+        parameter. If not provided, will be computed from Hessian.
         :param t_ppf: returns critical value as function of parameter
-        :param llr_tolerance: see bestfit
+        Use Wilks' theorem if omitted.
+        :param t_ppf_grad: return derivative of t_ppf
+        :param llr_tolerance: See bestfit
+        :param optimizer_kwargs: dict of additional arguments for optimizer
 
-        Return limit if kind='upper' or 'lower', returns interval if 'central'
+        Returns a float (for upper or lower limits)
+        or a 2-tuple of floats (for a central interval)
         """
+        if optimizer_kwargs is None:
+            optimizer_kwargs = dict()
+
         if bestfit is None:
-            # Determine global bestfit and global minimum -2lnL
+            # Determine global bestfit
             bestfit = self.bestfit(fix=fix, optimizer=optimizer)
 
-        # TODO, we could avoid this if bestfit returned grad and user passes it
-        ll_best, ll_best_grad = self.minus_ll(**bestfit)
-
-        slope_in_param = one_sigma = None   # keeps pycharm happy
-        if guess is None:
-            # Prepare to guess a limit ourselves, using our knowledge that the
-            # function is parabolic
-
-            # Estimate one sigma interval using parabolic approx
-            param_i = self.param_names.index(parameter)
-            one_sigma = cov_to_std(self.inverse_hessian(bestfit))[0][param_i]
-
-            # Remaining gradient in direction of parameter
-            slope_in_param = ll_best_grad[param_i].numpy()
-
         lower_bound = None
         if parameter.endswith('rate_multiplier'):
             lower_bound = fd.LOWER_RATE_MULTIPLIER_BOUND
 
         # Set (bound, critical_quantile) for the desired kind of limit
         if kind == 'upper':
-            limits = [dict(bound=(bestfit[parameter], None),
-                           crit=confidence_level,
-                           direction=1,
-                           guess=guess)]
+            requested_limits = [
+                dict(bound=(bestfit[parameter], None),
+                     crit=confidence_level,
+                     direction=1,
+                     guess=guess)]
         elif kind == 'lower':
```

Comment:
For 'lower' and 'central' we might need to check the bestfit value:
```assert lower_bound < bestfit[parameter]```

Reply:
Let's do bounds checks for the guesses, if the guess is in the bounds it's up to the minimizer. The tfp minimizer won't give results within the bounds as it ignores bounds.